### PR TITLE
[#158209998] Merge PagoPaAPI + PagoPa TS

### PR DESCRIPTION
**Pull request by LsKoder**
Merge with PagoPaAPI Project.
Generate types, client and server using PagoPA WSDL

### CODE REVIEW

**Comment by digitalcitizenship** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  Please include a Pivotal story at the beginning of the PR title (see below).
  </td>
    </tr>
  </tbody>
</table>



  Example of PR titles that include pivotal stories:

  * single story: `[#123456] my PR title`
  * multiple stories: `[#123456,#123457,#123458] my PR title`

    

New dependencies added: <a href='https://www.npmjs.com/package/apiconnect-wsdl'><code>apiconnect-wsdl</code></a>, <a href='https://www.npmjs.com/package/fs'><code>fs</code></a> and <a href='https://www.npmjs.com/package/js-yaml'><code>js-yaml</code></a>.


<h2><a href="https://developer.ibm.com/apiconnect">apiconnect-wsdl</a></h2>
<p>Author: Unknown</p>
<p>Description: API Connect WSDL</p>
<p>Homepage: <a href="https://developer.ibm.com/apiconnect">https://developer.ibm.com/apiconnect</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>about 2 years ago</td></tr><tr><td>Last Updated</td><td>13 days ago</td></tr><tr><td>License</td><td>SEE LICENSE IN LICENSE.md</td></tr><tr><td>Maintainers</td><td>13</td></tr><tr><td>Releases</td><td>25</td></tr><tr><td>Direct Dependencies</td><td><a href='https://www.npmjs.com/package/eslint'><code>eslint</code></a>, <a href='https://www.npmjs.com/package/iconv-lite'><code>iconv-lite</code></a>, <a href='https://www.npmjs.com/package/js-yaml'><code>js-yaml</code></a>, <a href='https://www.npmjs.com/package/jszip'><code>jszip</code></a>, <a href='https://www.npmjs.com/package/lodash'><code>lodash</code></a>, <a href='https://www.npmjs.com/package/q'><code>q</code></a>, <a href='https://www.npmjs.com/package/soap'><code>soap</code></a>, <a href='https://www.npmjs.com/package/swagger-parser'><code>swagger-parser</code></a>, <a href='https://www.npmjs.com/package/xmldom'><code>xmldom</code></a>, <a href='https://www.npmjs.com/package/yauzl'><code>yauzl</code></a> and <a href='https://www.npmjs.com/package/swagger2openapi'><code>swagger2openapi</code></a></td></tr>
</table>

<details>
<summary><code>README</code></summary></br>

# API Connect WSDL

## Getting Started
To get up and running, just add a require:

```javascript
var apicWsdl = require("apiconnect-wsdl");
```

## Test

Simply run npm install.  This will run 200+ tests and 1000+ asserts:
```javascript
  npm install
```
More complicated testing is described in the [general docs](docs/HOME.md)

## Links

 - Build and Release information is located in the [release docs](docs/RELEASE.md)  Note this module is used for v.Next and v5x

 - A private cli (wpcli) is available to execute functions. See  [general docs](docs/HOME.md)

 - If you have questions about WSDL, SOAP, generation see the [support docs](docs/SUPPORT.md) and the [matrix](docs/DETAILS.md)

 - The primary apis are listed below.  Older apis are located in [legacy apis](docs/OLDAPIS.md)

## Primary Apis

### create

```javascript
/**
 * Create an openapi from wsdl
 * @param {Buffer or String file location or String url} wsdl or zip content
 * @param {String} serviceName name of the wsdl service
 * @param {String} wsdlId id to use in the generated swagger.  Often this is the filename
 * @param options
 *     type: 'wsdl' (default) or 'wsdl-to-rest'
 *     openapiVersion: '2.0' (default) or '3.0'
 *     allowExtraFiles: false (default) or true
 *     defaults: object to merge into openapi after generation.  Same shape as openapi
 *     auth: auth object if wsdl is protected url
 *     mapOptions: setting to use for autogenerated assembly map. The default is no map.options
 *         mapOptions.includeEmptyXMLElements
 *         mapOptions.namespaceInheritance
 *         mapOptions.inlineNamespaces
 * @return {
 *              openapi: <openapi>
 *              statistics: { various stats}
 *              messages: [ {message: <info message>}* ]
 *         }
 */
function createOpenApi(wsdl, serviceName, wsdlId, options)
```
Creates an openapi from a wsdl.
  - The wsdl can be a Buffer containing the wsdl/zip contents or a url/filepath.
  - The serviceName is the name of the service within the wsdl for this created api.
  - The openapiversion is 2.0 or the new 3.0 version
  - The type is 'wsdl' (for wsdl proxy) or 'wsdl-to-rest' (SOAP->REST)
  - The create api has its own defaults for various fields.  You can provide your own overrides.

```javascript
  let result = await.apicWsdl.create(buffer, serviceName, filename,
	 {
     type: 'wsdl',
     openapiVersion: version,
     defaults: {
       securityDefinitions: {
         clientIdHeader: {
          type: 'apiKey',
          in: 'header',
          name: 'X-IBM-Client-Id',
         }
       },
       security: [ {
         clientIdHeader: []
       } ],
       'x-ibm-configuration': {
          cors: {
            enabled: false
          },
       }
     }
	 });
  // The openapi is within result.openapi

```

### validateWSDL

```javascript
/**
* Validates wsdl and rejects with error containing messages.
* @error { "message": <combined message>
*          messages[ {message: <message>}* ]
*        }
* @param locationOrContent location of WSDL or Buffer
* @param options auth and other options.
* @returns promise of serviceData
*/
function validateWSDL(locationOrContent, options);
```
The validateWSDL method will parse the wsdl/zip and perform some additional semantic validation of the wsdl.  Thrown errors list the individual messages

```javascript
  try {
    await.apicWsdl.validateWSDL(buffer);
	catch (err) {
		// Process messages
	}
```

### validateXSD

```javascript
/**
* Validates xsd and rejects with error containing messages.
* @error { "message": <combined message>
*          messages[ {message: <message>}* ]
*        }
* @param locationOrContent location of XSD or Buffer
* @param options auth and other options.
* @returns promise of serviceData
*/
function validateXSD(locationOrContent, options);
```
The validateWSDL method will parse the xsd/zip and perform some additional semantic validation of the xsd.  Thrown errors list the individual messages

```javascript
  try {
    await.apicWsdl.validateXSD(buffer);
	catch (err) {
		// Process messages
	}
```

### introspectWSDL

```javascript
/**
* Introspect wsdl and rejects with error containing messages.
* @error { "message": <combined message>
*          messages[ {message: <message>}* ]
*        }
* @param locationOrContent location of WSDL or Buffer
* @param options auth and other options.
* @returns promise of serviceData
*/
function introspectWSDL(locationOrContent, options)...
```
The introspectWSDL api is used to determine which services are in the
wsdl.

```javascript
Example Response
{ portTypes:
   { MyPortTypeType:
      [ { name: 'getStuff', description: undefined },
        { name: 'doStuff', description: undefined } ] },
  bindings:
   { MySOAPBinding:
      { type: 'MyPortType',
        operations: [ 'getStuff', 'doStuff' ] } },
  services:
   [ { service: 'MyService',
       filename: 'MyWSDL_1.0.0.wsdl',
       operations:
        [ { operation: 'getStuff' },
          { operation: 'doStuff' } ] } ] }
```

```javascript
  try {
    let serviceData = await.apicWsdl.introspectWSDL(buffer);
	catch (err) {
		// Process serviceData
	}
```

### addTargetOpenApi

```javascript
//**
* addTargetOpenApi
* @param openApi - existing openApi
* @param wsdl - Buffer containing the wsdl/zip or location of the wsdl/zip
* @param wsdlServiceName - Service
* @return promise openApi
*/
function addTargetOpenApi(openapi, wsdl, wsdlServiceName) ..
```
The addTargetOpenApi api is used to add a target service into an existing openapi.
  - The openapi is the existing openapi (version 2.0 or 3.0)
	- The wsdl is a Buffer, wsdl/zip file location or URL
	- The wsdl service is the name of the target service in the wsdl
	- returns an openapi with the embedded target service.


```javascript
    let updated_openapi = await.apicWsdl.addTargetOpenApi(openapi, wsdl, service);
```

### addXSDToTargetOpenApi

```javascript
/**
* addXSDToTargetOpenApi
* @param openApi - existing openApi
* @param xsd - Buffer containing the xsd or location of the xsd
* @param wsdlServiceName - Service
* @return promise openApi
*/
function addTargetOpenApi(openapi, xsd, wsdlServiceName) ..
```
The addXSDToTargetOpenApi api is used to add xsd definitions to a target service within an existing openapi.
  - The openapi is the existing openapi (version 2.0 or 3.0)
	- The xsd is a Buffer, xsd/zip file location or URL
	- The wsdl service is the name of the target service in the wsdl
	- returns an openapi with the embedded target service.


```javascript
    let updated_openapi = await.apicWsdl.addXSDToTargetOpenApi(openapi, xsd, service);
```

### injectServiceEndpointsIntoWSDLorZIP
```javascript
/**
* @param inContent wsdl or zip content (Buffer or String)
* @param serviceEndpoints single or array of endpoint strings
* @param serviceName the wsdl-definition.service string.
* @return (Promise)
*   outContent: wsdl or zip content in a Buffer
*   filename: if zip mode, this is the full name of first wsdl modified within the zip
*   stringContent: in zip mode, this is the string content of filename.
*                  in wsdl mode, this is the string content of the wsdl file
*/
function injectServiceEndpointsIntoWSDLorZIP(inContent, serviceEndpoints, serviceName)..
```
Use this method to add endpoints into a wsdl/zip file stored on apim.
This is used dynamically in the portal when a wsdl is requested.


</details>




<h2><a href="https://github.com/npm/security-holder#readme">fs</a></h2>
<p>Author: Unknown</p>
<p>Description: This package name is not currently in use, but was formerly occupied by another package. To avoid malicious use, npm is hanging on to the package name, but loosely, and we'll probably give it to you if you want it.</p>
<p>Homepage: <a href="https://github.com/npm/security-holder#readme">https://github.com/npm/security-holder#readme</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>about 4 years ago</td></tr><tr><td>Last Updated</td><td>4 months ago</td></tr><tr><td>License</td><td>ISC</td></tr><tr><td>Maintainers</td><td>1</td></tr><tr><td>Releases</td><td>3</td></tr>
</table>

<details>
<summary><code>README</code></summary></br>

# Security holding package

This package name is not currently in use, but was formerly occupied
by another package. To avoid malicious use, npm is hanging on to the
package name, but loosely, and we'll probably give it to you if you
want it.

You may adopt this package by contacting support@npmjs.com and
requesting the name.


</details>




<h2><a href="https://github.com/nodeca/js-yaml">js-yaml</a></h2>
<p>Author: Vladimir Zapparov</p>
<p>Description: YAML 1.2 parser and serializer</p>
<p>Homepage: <a href="https://github.com/nodeca/js-yaml">https://github.com/nodeca/js-yaml</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>over 6 years ago</td></tr><tr><td>Last Updated</td><td>about 1 month ago</td></tr><tr><td>License</td><td>MIT</td></tr><tr><td>Maintainers</td><td>1</td></tr><tr><td>Releases</td><td>64</td></tr><tr><td>Direct Dependencies</td><td><a href='https://www.npmjs.com/package/argparse'><code>argparse</code></a> and <a href='https://www.npmjs.com/package/esprima'><code>esprima</code></a></td></tr><tr><td>Keywords</td><td>yaml, parser, serializer and pyyaml</td></tr>
</table>
This README is too long to show.

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: package.json [hunk 1]
```diff
@@ -4,12 +4,14 @@
   "description": "Proxy that handles interactions with pagoPA",
   "main": "index.ts",
   "repository": "https://github.com/teamdigitale/italia-pagopa-proxy",
-  "author": "Sergio Manetta <sergio.manetta@it.ibm.com>",
+  "author": "Federico Feroldi <federico@teamdigitale.governo.it>",
```
Comment (cloudify) [inline package.json]: the proper way would be to have an `AUTHORS` file, add both of us and put "See AUTHORS" here

File: package.json [hunk 2]
```diff
@@ -4,12 +4,14 @@
...
   "license": "MIT",
   "scripts": {
-    "build": "npm-run-all -s generate:proxy-api-models lint build:tsc",
+    "build": "npm-run-all -s generate:proxy-api-models generate:yaml-from-wsdl generate:ts-from-yaml lint build:tsc && cp -r src/wsdl dist/wsdl",
     "build:tsc": "tsc",
-    "clean": "rm -r ./dist",
+    "clean": "rimraf dist && rm -f pagopa_api/yaml-file/*.yaml && rm -f src/types/pagopa_api/yaml-to-ts/*.ts",
+    "generate:ts-from-yaml": "rm -f src/types/pagopa_api/yaml-to-ts/*.ts && rm -f src/types/yaml-to-ts/*.ts && gen-api-models --strict 0 --api-spec ./pagopa_api/yaml-file/FespCdService.yaml --out-dir src/types/pagopa_api/yaml-to-ts/ && gen-api-models --strict 0 --api-spec ./pagopa_api/yaml-file/PagamentiTelematiciPspNodoservice.yaml --out-dir src/types/pagopa_api/yaml-to-ts/",
```
Comment (cloudify) [inline package.json]: I would perhaps split this command in multiple sub commands an call them with `npm-run-all`

File: pagopa_api/YamlGenerator.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
+/**
```
Comment (cloudify) [inline pagopa_api/YamlGenerator.js]: pls convert this to typescript

File: src/App.ts [hunk 4]
```diff
@@ -131,25 +129,39 @@ function setRestfulRoutes(
...
-        {
-          host: config.REDIS_DB.HOST,
-          port: config.REDIS_DB.PORT
-        }
-      ]
-    }) as redis.RedisClient;
-  }
-  logger.debug("Creating a REDIS client...");
-  return redis.createClient(config.REDIS_DB.HOST);
+  const redisClient = (() => {
```
Comment (cloudify) [inline src/App.ts]: what does this change have to do with merging the pagopa-api code?

File: package.json [hunk 5]
```diff
@@ -31,12 +33,15 @@
     "@types/supertest": "^2.0.4",
     "@types/uuid": "^3.4.3",
     "@types/winston": "^2.3.9",
+    "apiconnect-wsdl":"^1.6.3",
     "danger": "^3.7.16",
     "danger-plugin-digitalcitizenship": "^0.3.1",
+    "fs": "^0.0.1-security",
```
Comment (cloudify) [inline package.json]: this is a mistake I guess, you want `@types/node` instead

File: src/utils/SoapUtils.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+export function createClient<T>(
+  wsdlUri: string,
+  options: soap.IOptions
+): Promise<soap.Client & T> {
+  return new Promise((resolve, reject) => {
+    soap.createClient(wsdlUri, options, (err, client) => {
+      if (err) {
+        reject(err);
+      } else {
+        resolve(client as soap.Client & T); // tslint:disable-line
```
Comment (cloudify) [inline src/utils/SoapUtils.ts]: specify the tslint rule you want to disable

File: src/utils/PaymentsConverter.ts [hunk 7]
```diff
@@ -209,24 +217,43 @@ function getCodiceIdRpt(rptId: RptId): PPTPortTypes.IcodiceIdRPT {
...
+}
+
+/**
+ * Provide CodiceContestoPagamento element for PagoPa API
+ * parsing the CodiceContestoPagamento element provided by PagoPa Backend
+ */
+function getCodiceContestoPagamentoForPagoPaApi(
+  codiceContestoPagamento: CodiceContestoPagamento
+): stText35_ppt {
+  return stText35_ppt.decode(codiceContestoPagamento).value as stText35_ppt; // tslint:disable-line
```
Comment (cloudify) [inline src/utils/PaymentsConverter.ts]: specify the tslint rule you want to disable

**Pull request merged**